src/maker/server.rs

```diff
@@ -0,0 +1,561 @@
+                    let mut proof = maker.highest_fidelity_proof.write()?;
+                    *proof = Some(highest_proof);
+                    log::info!("[{}] Syncing and saving wallet data", maker.config.port);
+                    maker.get_wallet().write()?.sync()?;
+                    maker.get_wallet().read()?.save_to_disk()?;
+                    log::info!("[{}] Sync and save successful", maker.config.port);
+                    break;
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Keep checking if the Bitcoin Core RPC connection is live. Sets the global `accepting_client` flag as per RPC connection status.
+///
+/// This will not block. Once Core RPC connection is live, accepting_client will set as `true` again.
+fn check_connection_with_core(
+    maker: Arc<Maker>,
+    accepting_clients: Arc<Mutex<bool>>,
+) -> Result<(), MakerError> {
+    let mut rpc_ping_success = false;
+    while !*maker.shutdown.read()? {
+        // If connection is disrupted keep trying at heart_beat_interval (3 sec).
+        // If connection is live, keep tring at rpc_ping_interval (60 sec).
+        match rpc_ping_success {
+            true => {
+                sleep(Duration::from_secs(maker.config.rpc_ping_interval_secs));
+            }
+            false => {
+                sleep(Duration::from_secs(maker.config.heart_beat_interval_secs));
+            }
+        }
+        if let Err(e) = maker.wallet.read()?.rpc.get_blockchain_info() {
+            log::info!(
+                "[{}] RPC Connection failed. Reattempting {}",
+                maker.config.port,
+                e
+            );
+            rpc_ping_success = false;
+        } else {
+            rpc_ping_success = true;
+        }
+        let mut mutex = accepting_clients.lock()?;
+        *mutex = rpc_ping_success;
+    }
+
+    Ok(())
+}
+
+/// Handle a single client connection.
+fn handle_client(
+    maker: Arc<Maker>,
+    stream: &mut TcpStream,
+    client_addr: SocketAddr,
+) -> Result<(), MakerError> {
+    stream.set_read_timeout(Some(Duration::from_secs(
+        maker.config.idle_connection_timeout,
+    )))?;
+
+    stream.set_write_timeout(Some(Duration::from_secs(
+        maker.config.idle_connection_timeout,
+    )))?;
+
+    let mut connection_state = ConnectionState::default();
+
+    while !*maker.shutdown.read()? {
+        let mut taker_msg_bytes = Vec::new();
+        match read_message(stream) {
+            Ok(b) => taker_msg_bytes = b,
+            Err(e) => {
+                if let NetError::IO(e) = e {
+                    if e.kind() == ErrorKind::UnexpectedEof {
+                        continue;
+                    }
+                }
+            }
+        }
+        let taker_msg: TakerToMakerMessage = serde_cbor::from_slice(&taker_msg_bytes)?;
+        log::info!("[{}]  <=== {}", maker.config.port, taker_msg);
+
+        let reply = handle_message(
+            &maker.clone(),
```

Comment:
Is it necessary to clone here instead of simply passing it as a reference?
```suggestion
            &maker,
```

Reply:
Good catch. Changed.